extensions-core/azure-extensions/src/main/java/org/apache/druid/data/input/azure/AzureStorageAccountInputSource.java

```diff
@@ -0,0 +1,253 @@
+      @JacksonInject AzureInputDataConfig inputDataConfig,
+      @JacksonInject AzureAccountConfig azureAccountConfig,
+      @JsonProperty("uris") @Nullable List<URI> uris,
+      @JsonProperty("prefixes") @Nullable List<URI> prefixes,
+      @JsonProperty("objects") @Nullable List<CloudObjectLocation> objects,
+      @JsonProperty("objectGlob") @Nullable String objectGlob,
+      @JsonProperty("properties") @Nullable AzureInputSourceConfig azureInputSourceConfig,
+      @JsonProperty(SYSTEM_FIELDS_PROPERTY) @Nullable SystemFields systemFields
+  )
+  {
+    super(SCHEME, uris, prefixes, objects, objectGlob, systemFields);
+    this.entityFactory = Preconditions.checkNotNull(entityFactory, "AzureEntityFactory");
+    this.azureCloudBlobIterableFactory = Preconditions.checkNotNull(
+        azureCloudBlobIterableFactory,
+        "AzureCloudBlobIterableFactory"
+    );
+    this.inputDataConfig = Preconditions.checkNotNull(inputDataConfig, "AzureInputDataConfig");
+    this.azureInputSourceConfig = azureInputSourceConfig;
+    this.azureAccountConfig = azureAccountConfig;
+    this.azureIngestClientFactory = new AzureIngestClientFactory(
+        azureAccountConfig,
+        azureInputSourceConfig
+    );
+  }
+
+  @JsonIgnore
+  @Nonnull
+  @Override
+  public Set<String> getTypes()
+  {
+    return Collections.singleton(SCHEME);
+  }
+
+  @Override
+  public SplittableInputSource<List<CloudObjectLocation>> withSplit(InputSplit<List<CloudObjectLocation>> split)
+  {
+    return new AzureStorageAccountInputSource(
+        entityFactory,
+        azureCloudBlobIterableFactory,
+        inputDataConfig,
+        azureAccountConfig,
+        null,
+        null,
+        split.get(),
+        getObjectGlob(),
+        azureInputSourceConfig,
+        systemFields
+    );
+  }
+
+  @Override
+  public Object getSystemFieldValue(InputEntity entity, SystemField field)
+  {
+    final AzureEntity azureEntity = (AzureEntity) entity;
+
+    switch (field) {
+      case URI:
+        return azureEntity.getUri().toString();
+      case BUCKET:
+        return azureEntity.getLocation().getBucket();
+      case PATH:
+        return azureEntity.getLocation().getPath();
+      default:
+        return null;
+    }
+  }
+
+  @Override
+  protected AzureEntity createEntity(CloudObjectLocation location)
+  {
+    return entityFactory.create(
+        location,
+        new AzureStorage(azureIngestClientFactory, location.getBucket()),
+        SCHEME
+    );
+  }
+
+  @Override
+  protected CloudObjectSplitWidget getSplitWidget()
+  {
+    class SplitWidget implements CloudObjectSplitWidget
+    {
+      @Override
+      public Iterator<LocationWithSize> getDescriptorIteratorForPrefixes(List<URI> prefixes)
+      {
+        return Iterators.transform(
+            azureCloudBlobIterableFactory.create(prefixes, inputDataConfig.getMaxListingLength(), new AzureStorage(azureIngestClientFactory, null)).iterator(),
```

Comment:
Do you need to create a new `AzureStorage` instance for each prefix? Or can we instantiate one and reuse it for all?

Reply:
this is only creating 1 azureStorage instance for all the instances afaict

Reply:
ahh nvm, misread